keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+- If dual-stack endpoint address pairs are present, then each pair should be printed on its own line as show below.
+Example:
+```
+       kube-master# kubectl get endpoints
+       NAME            ENDPOINTS                             AGE
+       kubernetes      {[fd00::100]:6443,10.0.0.2:6643}      15m
+       nginx-service   {[fd00:db8:1::2]:80,192.168.1.3:80}   20m
+                       {[fd00:db8:2::2]:80,192.168.2.3:80}   20m
+       kube-master#
+```
+
+#### 'kubectl describe service' Command Display for Dual-Stack Backend Pods
+The 'kubectl describe service ...' command display should be changed so that:
+- IPv4/IPv6 endpoint address pairs appear within curly braces.
+- If dual-stack endpoint address pairs are present, then each pair should be printed on its own line as show below.
+Example:
+```
+       kube-master# kubectl describe service nginx-service
+       .
+       .
+       .
+       Endpoints:   {[fd00:db8:1::2]:80,192.168.1.3:80}
+                    {[fd00:db8:2::2]:80,192.168.2.3:80}
+       .
+       .
+       .
+       kube-master#
+```
+
+### kube-proxy Operation
+
+Kube-proxy will be modified to drive iptables and ip6tables in parallel. This will require the implementation of a second "proxier" interface in the Kube-Proxy server in order to modify and track changes to both tables. This is required in order to allow exposing services via both IPv4 and IPv6, e.g. using Kubernetes:
+  - NodePort
+  - ExternalIPs
+
+#### Kube-Proxy Startup Configuration Changes
+
+##### Multiple bind addresses configuration
+The existing "--bind-address" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) will be modified to support multiple IP addresses in a comma-separated list (rather than a single IP string).
+```
+  --bind-address  stringSlice   (IP addresses, in a comma separated list, Default: [0.0.0.0,])
+```
+Only the first address of each IP family will be used; all others will be ignored.
+
+##### Multiple cluster CIDRs configuration
+The existing "--cluster-cidr" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) will be modified to support multiple IP CIDRs in a comma-separated list (rather than a single IP CIDR).
+A new [kube-proxy configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) argument will be added to allow a user to specify multiple cluster CIDRs.
+```
+  --cluster-cidr  ipNetSlice   (IP CIDRs, in a comma separated list, Default: [])
+```
+Only the first CIDR for each IP family will be used; all others will be ignored.
+
+### IPVS Support and Operation
+
+Since IPVS functionality does not yet include IPv6 support (see [cloudnativelabs/kube-router Issue #307](https://github.com/cloudnativelabs/kube-router/issues/307)), support for IPVS functionality in a dual-stack cluster is considered a "nice-to-have" or stretch goal.
```

Comment:
I didn't realize teh IPVS kube-proxy doesn't do v6 at all - is that right?

@m1093782566 can you comment on this?  I am reluctant to GA without it